lib/Runtime/Base/FunctionExecutionStateMachine.h

```diff
@@ -15,50 +15,73 @@ namespace Js
 
         // Public Getters and Setters
         ExecutionMode GetExecutionMode() const;
-        void SetExecutionMode(ExecutionMode mode); // This should eventually become private
-        ExecutionMode GetDefaultInterpreterExecutionMode() const;
+        void SetDefaultInterpreterExecutionMode();
+        void SetAsmJsExecutionMode();
+
+        // Interpreter-related functions
         ExecutionMode GetInterpreterExecutionMode(const bool isPostBailout);
         bool IsInterpreterExecutionMode() const;
         uint16 GetProfiledIterations() const;
-
         uint32 GetInterpretedCount() const { return interpretedCount; }
         uint32 SetInterpretedCount(uint32 val) { return interpretedCount = val; }
         uint32 IncreaseInterpretedCount() { return interpretedCount++; }
         bool InterpretedSinceCallCountCollection() const;
         void CollectInterpretedCounts();
         void CommitExecutedIterations();
 
+        // JIT-relatedfunctions
         void SetIsSpeculativeJitCandidate();
         uint16 GetSimpleJitLimit() const { return simpleJitLimit; }
         void ResetSimpleJitLimit();
         uint16 GetSimpleJitExecutedIterations() const;
+        void SetSimpleJitCallCount(const uint16 simpleJitLimit) const;
         void SetFullJitThreshold(const uint16 newFullJitThreshold, const bool skipSimpleJit = false);
         uint16 GetFullJitThreshold() const { return fullJitThreshold; }
         void SetFullJitRequeueThreshold(const uint16 newFullJitRequeueThreshold);
-        void SetSimpleJitCallCount(const uint16 simpleJitLimit) const;
 
         // Transition functions
         bool TryTransitionToNextExecutionMode();
         void TryTransitionToNextInterpreterExecutionMode();
         bool TryTransitionToJitExecutionMode();
         void TransitionToSimpleJitExecutionMode();
         void TransitionToFullJitExecutionMode();
-
+        
+        // Debug functions
         void PrintLimits() const;
         void AssertIsInitialized() const;
 
-
     private:
+        // This enum creates a linear view of the progression of ExecutionModes, as
+        // described in the banner comment for TryTransitionToNextExecutionMode
+        enum ExecutionState
+        {
+            Interpreter,
```

Comment:
ExecutionModes.h had good comments about the different modes.  It would be nice to have the same.
Do we still need ExecutionMode?  Should it be removed?

Reply:
I was also thinking about do we still need ExecutionMode, but was on the fence with it.

---
In reply to: [145004520](https://github.com/Microsoft/ChakraCore/pull/3941#discussion_r145004520) [](ancestors = 145004520)

Reply:
So, I didn't add as much about the modes here because I didn't remove ExecutionModes. In the latest commit, I did add a comment saying to refer to Executionmodes.h for more details.

I wanted ExecutionState to remain an internal detail to the state machine and not be something that's exposed outside of it. If I remove ExecutionMode, I'll need to also update places throughout the code that uses this enum and also come up with a way to represent how some modes are repeated in the state diagram. I'd rather not change that, but won't push back if you say I should.